src/api/app/helpers/webui/staging/workflow_helper.rb

```diff
@@ -0,0 +1,83 @@
+module Webui::Staging::WorkflowHelper
+  def build_progress(staging_project)
+    total = 0
+    final = 0
+
+    staging_project.building_repositories.each do |r|
+      total += r[:tobuild] + r[:final]
+      final += r[:final]
+    end
+
+    return 100 if total == 0
+    # if we have building repositories, make sure we don't exceed 99
+    [final * 100 / total, 99].min
+  end
+
+  def review_progress(staging_project)
+    staged_requests_numbers = staging_project.staged_requests.pluck(:number)
+    total = Review.where(bs_request: staging_project.staged_requests).count
+    missing = staging_project.missing_reviews.count { |missing_review| staged_requests_numbers.include?(missing_review[:request]) }
+
+    100 - missing * 100 / total
+  end
+
+  def testing_progress(staging_project)
+    notdone = 0
+    allchecks = 0
+
+    # Note: The status_reports are defined via a has many through relation. Since within the
+    #       status report context the bs request relation is polymorphic, we have to call
+    #       includes with the polymorphic name ('checkable').
+    staging_project.status_reports.includes(:checkable).each do |report|
+      notdone += report.checks.where(state: 'pending').count
+      allchecks += report.checks.count + report.missing_checks.count
+    end
+
+    return 100 if allchecks == 0
+    100 - notdone * 100 / allchecks
+  end
+
+  def progress(staging_project)
+    case staging_project.overall_state
+    when :building
+      link_to project_monitor_url(staging_project.name) do
+        "#{build_progress(staging_project)} %"
+      end
+    when :review
+      "#{review_progress(staging_project)} %"
+    when :testing
+      "#{testing_progress(staging_project)} %"
+    end
+  end
+
+  def requests(staging_project)
+    requests_count = staging_project.classified_requests.count
+
+    return 'None' if requests_count == 0
+
+    requests_visible_by_default = 10
+    requests_links = staging_project.classified_requests.map do |request|
+      css = 'ready'
+      css = 'review' if request[:missing_reviews].present?
+      css = 'obsolete' if request[:state].in?(BsRequest::OBSOLETE_STATES)
+
+      link_to(request[:package], request_show_path(request[:number]), class: "request state-#{css}")
+    end
+
+    if requests_count <= requests_visible_by_default
+      return safe_join(requests_links)
+    end
+
+    output = safe_join(requests_links[0, requests_visible_by_default])
+
+    # Add requests links which aren't visible by default in a collapsible div controlled by a clickable icon
```

Comment:
@krauselukas and I have implemented this for the other column in this page in a completely different way. I think we should sync.

Reply:
I built this using the [Collapse component from Bootstrap](https://getbootstrap.com/docs/4.1/components/collapse/)

Reply:
We talked. Marking this comment as resolved.